### ISSUE TITLE
[NNBD] Type promotion fails for variable uses in anonymous functions

### ISSUE THREAD

**Issue by nex3** [dart-lang/sdk#45348]
The following function:

```dart
int foo({int? arg}) {
  arg = 0;
  return (() => arg)();
}
```

Produces this error:

```
  error • A value of type 'int?' can't be returned from the function 'foo' because it has a return type of 'int'. • test.dart:3:10 • return_of_invalid_type
```

Despite the fact that `arg` will clearly always be non-null when it's returned. There's not even a hypothetical risk that `arg` will be reassigned to `null` between the function being created and being invoked, since the non-nullable assignment appears before the function is created and no nullable assignments exist anywhere.

**Comment by stereotype441** [comment]
Yeah, this is a known limitation of type promotion.  When it sees an anonymous function, it makes a pretty conservative assumption about what might happen to any variables that anonymous function reads: it reasons that if the variable is *ever* written to, then it's possible that the variable's promotion will be invalidated by a write sometime between the time it's created and the time it's called.  So to be on the safe side, it invalidates the type promotion.

Obviously in this example, that's overly conservative, for three reasons: (a) because the anonymous function is called exactly once, immediately after it's created, (b) because the only assignment to `arg` happens *before* the anonymous function is created, and (c) because that assignment always assigns a non-null value.

I believe we have some other bugs in the language repo discussing similar issues.  I'll try to dig them up and link them here, so that we remember to think about this case when we think about future improvements to type promotion.

**Comment by stereotype441** [comment]
FWIW, you can work around this limitation by creating a temporary variable that's never written to:
```dart
int foo({int? arg}) {
  arg = 0;
  var tmp = arg;
  return (() => tmp)();
}
```

**Comment by nex3** [comment]
This is a pretty rough thing to have to work around—Dart heavily encourages the use of anonymous functions for things like collection manipulation, so added friction passing values in produces a lot of pain.

> Obviously in this example, that's overly conservative, for three reasons: (a) because the anonymous function is called exactly once, immediately after it's created, (b) because the only assignment to `arg` happens _before_ the anonymous function is created, and (c) because that assignment always assigns a non-null value.

For what it's worth, (a) here is basically a coincidence—in the actual code in which I'm seeing this issue, there's no way for the analyzer to know that the callback is only called once. But (b) and (c) are both usually true for these situations.
